src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+        self._layer_list = {i: [] for i in range(self._qwires + self._cwires)}
+
+        self._render_strs = {
+            "top_frame": ["  "] * (self._qwires + self._cwires),
+            "mid_frame": ["──"] * self._qwires + ["══"] * self._cwires,
+            "bot_frame": ["  "] * (self._qwires + self._cwires),
+        }
+
+    def _adjust_layer(self, wire_list: List[int], xskip: int):
+        """
+        Adjust the layers by filling the empty spaces with respective characters
+        """
+
+        for wire in wire_list:
+            self._render_strs["top_frame"][wire] += " " * (
+                xskip - len(self._render_strs["top_frame"][wire])
+            )
+            self._render_strs["bot_frame"][wire] += " " * (
+                xskip - len(self._render_strs["bot_frame"][wire])
+            )
+            if wire < self._qwires:
+                self._render_strs["mid_frame"][wire] += "─" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+            else:
+                self._render_strs["mid_frame"][wire] += "═" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
+
+    def _extend_line(self):
+        """
+        Extend all the wires to the same length
+        """
+
+        wire_len = [sum(wire) for wire in self._layer_list.values()]
+        max_len = max(wire_len) + self.style.end_wire_ext
+
+        for i, length in enumerate(wire_len):
+            if length < max_len:
+                diff = max_len - length
+                if i < self._qwires:
+                    self._render_strs["mid_frame"][i] += "─" * diff
+                else:
+                    self._render_strs["mid_frame"][i] += "═" * diff
+
+    def _draw_singleq_gate(self, gate_name: str):
+        """
+        Draw a single qubit gate
+
+        Parameters
+        ----------
+        gate : Gate
+            The gate to be drawn.
+
+        Returns
+        -------
+        tuple
+            The parts of the gate to be drawn. The parts are the top_frame, mid_frame, and bot_frame.
+
+        int
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_name))
```

Comment:
From the default Style, `gate_pad=0.05`, however here I believe we need an integer here? `ceil` is a bit weird. It will cause weird behaviour. I think we should just have a different default value. It seems that it would be nice to summarize the default value somewhere. This is related to the other comment of mine yesterday. Maybe have a separate default style for each renderer?